Add bip340 schnorr

Tibo-lg opened the pull request:
As BIP 340 support was [merged in bitcoin core](https://github.com/bitcoin/bitcoin/pull/19944) and that I had a branch of rust-secp256k1 with BIP 340 support I thought maybe now was a good time to try to get it merged.

Changes:
* Update the depend folder to the latest master
* Update `secp256k1-sys` to add Schnorr and extra keys modules, as well as the bindings for the related functions
* Add `schnorrsig` module in the `rust-secp256k1`

One thing I am unsure is whether types from the extra keys module should be used into the public interfaces (x-only pubkeys and keypairs). For now I implemented to just use `PublicKey` and `SecretKey` in public interface, but I can change it.

elichai commented:
I think we should have a discussion on what API do we want to use here and how.
IMHO we should *really* discourage people from using the same keys for schnorr and ecdsa, do we do this by having even a separate privkey struct for schnorr? idk. 
but at least I think we shouldn't allow verifying schnorr from an ECDSA pubkey without explicitly converting between the keys.


Maybe we can schedule an IRC call on this sometime? @apoelstra @real-or-random

Tibo-lg commented:
I'm happy to add separate type if that's what you think it's best (and if you're ok with it). I tried to keep it minimal for now. But that kind of feedback is really helpful as I'm maintaining a separate branch for DLC things and the less I diverge from mainstream the easier it is to maintain. Anyway keep me updated!

Tibo-lg commented:
Updated to add `SchnorrPublicKey` struct (which corresponds to `XOnlyPublicKey` at the sys module level) as it makes quite some things easier for me and simplifies the code a little bit.

apoelstra commented on src/schnorrsig.rs:
```diff
@@ -0,0 +1,471 @@
+
+/// A Secp256k1 public key, used for verification of signatures
+#[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub struct SchnorrPublicKey(ffi::XOnlyPublicKey);
```
This can just be called `PublicKey`, it's already in the `bip340` module, so users can refer to it as `bip340::PublicKey`

Tibo-lg replied:
Done

apoelstra commented on src/lib.rs:
```diff
@@ -169,6 +169,7 @@ mod context;
 pub mod constants;
 pub mod ecdh;
 pub mod key;
+pub mod schnorrsig;
```
I'd like the module to be called `bip340` rather than `Schnorr`. Both as a dig at Schnorr for patenting things, and also to be clear that this is a specific Schnorr-based signature scheme specified in BIP340.

Tibo-lg replied:
Schnorr has been removed from this module :).  Only mentions left are in the sys module as the method calls of secp use it.

apoelstra commented:
+1 to having a separate `SecretKey` type in the bip340 module

apoelstra commented on src/schnorrsig.rs:
```diff
@@ -0,0 +1,471 @@
+use {Message, Signing};
+
+/// Represents a schnorr signature.
+pub struct SchnorrSignature([u8; constants::SCHNORR_SIGNATURE_SIZE]);
```
Similarly we can drop the prefix from this type

Tibo-lg replied:
Done

Tibo-lg commented:
* Changed names from Schnorr to BIP340 wherever possible
* Made macros for implementing SecretKey functions and the serde serialization for PublicKey to reduce code duplication. If you don't like that I can replace it with simple copy paste but it felt nicer that way. There might be room for adding some PublicKey functions to macros, though not all due to the uncompressed version for ECDSA.
* Added `SecretKey` in BIP340 module (which has the exact same implementation as the one in the `key` module through the above mentioned macro).

Also wanted to mention a couple of things I haven't handled:
* tweak add for `bip340::PublicKey`
* I just put `unimplemented!()` in the `fuzz_dummy` module for the functions I added
I think those can be handled later but if let me know if they are essential.

apoelstra commented:
I think it's fine to macro-ize the serde stuff. Can you use the same macro to impl serde on the `bip340::Signature` type?

As for the secret key stuff, it'd be better to move the code into a new `key.rs.in` file and then `include!("key.rs.in"); // define SecretKey` in both files. Rather than using `macro_rules` which adds extra indentation and tends to result in confusing compiler error messages.

real-or-random commented:
I think bip340 is fine but it may be a good idea to use the same module names as upstream.

Tibo-lg commented:
@apoelstra 
> I think it's fine to macro-ize the serde stuff. Can you use the same macro to impl serde on the bip340::Signature type?

I couldn't use the same one as `PublicKey` but used the one that was already there for `SecretKey` (it had a harcoded `SecretKey` I assumed that was a typo: https://github.com/rust-bitcoin/rust-secp256k1/pull/237/commits/873537879085503085796ca95675e360b9d30a91#diff-e1576943a5907b15388ab726037ebdb3L70).

> As for the secret key stuff, it'd be better to move the code into a new key.rs.in file and then include!("key.rs.in"); // define SecretKey in both files. Rather than using macro_rules which adds extra indentation and tends to result in confusing compiler error messages.

Done. That's neat didn't know about this feature.

Pipeline is failing now but AFAICT it seems to be some issue with rustup installation of nightly version.

apoelstra commented on secp256k1-sys/src/lib.rs:
```diff
@@ -232,76 +289,76 @@ extern "C" {
                                          tweak: *const c_uchar)
                                          -> c_int;
 
-    #[deprecated(since = "0.2.0",note = "Please use the secp256k1_ec_seckey_tweak_mul function instead")]
```
Did you change this on purpose?

Tibo-lg replied:
Sorry not sure what I've done there, fixed.

apoelstra commented:
@Tibo-lg can you change the first commit to include all the 0_2_0→0_2_1 mappings in the .rs files? otherwise it doesn't build.

apoelstra commented:
Can you also reduce the indentation level in `key.rs.in`, and also move
```
/// Secret 256-bit key used as `x` in an ECDSA signature
pub struct SecretKey([u8; constants::SECRET_KEY_SIZE]);
```
(I guess, update the comment to not say ECDSA) into the file? It makes the diff easier to read.

Tibo-lg commented:
> @Tibo-lg can you change the first commit to include all the 0_2_0→0_2_1 mappings in the .rs files? otherwise it doesn't build.

Done. I ran a `cargo build && cargo test` on each commit and it seems to pass. 

Also do you want to bump the cargo versions as part of the changes or you'll do that later?

> Can you also reduce the indentation level in key.rs.in, and also move
>
>/// Secret 256-bit key used as `x` in an ECDSA signature
>pub struct SecretKey([u8; constants::SECRET_KEY_SIZE]);
>(I guess, update the comment to not say ECDSA) into the file? It makes the diff easier to read.

Done.

apoelstra commented:
Sorry to nit again, but I realize that we've messed up the linking versions here..

Can you
1. Change to version `0_3_1` rather than `0_2_1`
2. Increase the secp256k1-sys/Cargo.toml version from `0.3.0` to `0.3.1`?

Tibo-lg commented:
> Sorry to nit again, but I realize that we've messed up the linking versions here..

No problem

> 1. Change to version 0_3_1 rather than 0_2_1

Done: https://github.com/rust-bitcoin/rust-secp256k1/pull/237/commits/cfb565126a6e7e2091e7215e07f325aefa9c9337#diff-e3309566a1a56f6f7ace19f2697d700bR15

> 2. Increase the secp256k1-sys/Cargo.toml version from 0.3.0 to 0.3.1?

Done: https://github.com/rust-bitcoin/rust-secp256k1/pull/237/commits/bf696456fc34210a79ce2911bf0adf17272eea7d#diff-e3309566a1a56f6f7ace19f2697d700bR3

elichai commented on src/bip340.rs:
```diff
@@ -0,0 +1,476 @@
+
+impl<C: Signing> Secp256k1<C> {
+    /// Create a BIP340 signature.
+    pub fn bip340_sign(
+        &self,
+        msg: &Message,
+        sk: &SecretKey,
+        aux_rand: &[u8; 32],
```
IMO this should be splitted to 2 functions:
```rust
bip340_sign(&self, &Message, &SecretKey)
bip340_sign_randomize(&self, &Message, &SecretKey, rng: impl Rng + CryptoRng)
```
or something like that

Tibo-lg replied:
I agree it's nice to have some convenience functions. I defined three now:
```rust
bip340_sign(&self, msg: &Message, sk: &SecretKey)
```
Internally uses `OsRng` to generate the auxiliary random data. Requires `rand-std` feature (I hope I understood the purpose of this feature correctly)

```rust
bip340_sign_with_rng<R: Rng + CryptoRng>(
        &self,
        msg: &Message,
        sk: &SecretKey,
        rng: &mut R,
    )
```
Uses the given rng to generate the auxiliary random data. Requires `rand` feature.

```rust
bip340_sign_with_aux_rand(
        &self,
        msg: &Message,
        sk: &SecretKey,
        aux_rand: &[u8; 32],
    )
```
Uses the provided auxiliary random data, doesn't require any special feature. I think it's nice to keep that as it makes it possible to create BIP340 signatures without compiling with "rand" (also practical for testing, but could make it crate pub if you think it's better not to have it public).

I hope that matches with what you had in mind.

elichai replied:
I would just like to point out that this is diverging from out usual API (not saying that's necessarily bad), our usuall API takes an `Rng` for random, and in the test we use fake Rng's. 
Personally I don't mind too much. @apoelstra what do you think?

Tibo-lg replied:
Ah maybe I have misunderstood how you wanted the function `bip340_sign`. As you didn't put any rng parameter there nor auxiliary random data I assumed you wanted it to be instantiated inside the function, but how were you thinking about it? I can just drop the `bip340_sign` function if instantiating an rng within the library is not desired.

> in the test we use fake Rng's.

I saw quite some tests using `thread_rng` for key pair generation so I thought I would use that as well, but if you think it's better to have something deterministic I can replace maybe with `DumbRng`?

elichai replied:
> I saw quite some tests using thread_rng for key pair generation so I thought I would use that as well, but if you think it's better to have something deterministic I can replace maybe with DumbRng?

That depends on what you want to test, you asked for a way to test with a specific aux_rand

> Ah maybe I have misunderstood how you wanted the function bip340_sign. As you didn't put any rng parameter there nor auxiliary random data I assumed you wanted it to be instantiated inside the function, but how were you thinking about it?

I was thinking about not using aux rand there, and passing NULL

Tibo-lg replied:
> That depends on what you want to test, you asked for a way to test with a specific aux_rand

Got you, indeed I can get rid of the `_with_aux_rand` version if you don't think it's useful for users and replace the tests with a dumb rng.

elichai commented on src/bip340.rs:
```diff
@@ -0,0 +1,524 @@
+    /// Create a bip340 signature using OsRng to generate the auxiliary random
+    /// data. Requires compilation with "rand-std" feature.
+    #[cfg(any(test, feature = "rand-std"))]
+    pub fn bip340_sign(&self, msg: &Message, sk: &SecretKey) -> Result<Signature, Error> {
+        let mut rng = OsRng::new().expect("OsRng");
+        self.bip340_sign_with_rng(msg, sk, &mut rng)
+    }
```
This should call 
```rust
                ffi::secp256k1_schnorrsig_sign(
                    self.ctx,
                    sig.as_mut_c_ptr(),
                    msg.as_c_ptr(),
                    &keypair,
                    ffi::secp256k1_nonce_function_bip340,
                    ptr::null(),
                )
```

Tibo-lg replied:
Oh ok now I understand what you want. But is it a good idea to have the most obvious function name `bip340_sign` for this then? My understanding of bip340 is that using random auxiliary data should be encouraged, but here I'm afraid that this will become the default choice for users of the library. Maybe then it should be:
```rust
bip340_sign<R: Rng + CryptoRng>(
        &self,
        msg: &Message,
        sk: &SecretKey,
        rng: &mut R,
    )
```
```rust
bip340_sign_with_aux_rand(
        &self,
        msg: &Message,
        sk: &SecretKey,
        aux_rand: &[u8; 32],
    )
```
```rust 
bip340_sign_no_aux_rand(
        &self,
        msg: &Message,
        sk: &SecretKey
    )
```

What do you think?

elichai replied:
I'm ok with these, just the 2nd one I'm ~0 on it (not against but not for)

Tibo-lg replied:
Updated as mentioned above (with some refactor to avoid duplicate code).

Also I skimmed through [the comments](https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1_schnorrsig.h#L52) in the upstream repo and they mention that passing null "does not produce BIP-340 compliant signatures" so just wanted to confirm that you really want to give this option.

real-or-random replied:
Hm, I think the default `sign` function should simply pass `thread_rng` here (at least if we have `std`). I'd be fine with adding yet another function `bip340_sign_with_aux_rng`. We could maybe drop `bip340_sign_with_aux_rand` instead, I'm not sure if this is helpful in Rust. Does this make sense?

elichai replied:
@real-or-random You want to provide 2 separate `sign` functions depending on if `rand-std` is enabled?

real-or-random replied:
what I had in mind is something along these lines:
 - if  `rand-std` enabled: `_sign`
 - always: `_sign_with_aux_rng` and `_sign_without_aux_rand` (or `sign_without_aux_rng`)

real-or-random replied:
Okay, now that I looked at the current version, I think having `_sign_with_aux_rand` is good to have because it does not need the rand module.

So I believe we should have
 * if `rand-std` enabled: `_sign`
 * if `rand` enabled: `_sign_with_aux_rng` (this should really be `_sign_with_aux_rng` and only `_sign_with_prg` because the latter sounds like we don't use deterministic nonces...)
 * always: `_sign_with_aux_rand` and `_sign_without_aux_rand`.


But let's first agree on this before you need to go back and forth with that as well. @elichai @apoelstra Does this sound good? These are four functions. May be overkill but I think all have their reason to exist.

real-or-random commented:
> I think bip340 is fine but it may be a good idea to use the same module names as upstream.

Noone replied here.

apoelstra commented:
Agreed, let's match upstream module names ... we can use "bip340" for the support in rust-bitcoin.

real-or-random commented:
By the way, @Tibo-lg, sorry for going back and forth and the module name...

Tibo-lg commented:
@real-or-random no problem. I updated the module name to `schnorrsig` now, but I kept `bip340` as function name prefix. Personally I think it might be better to be consistent and use `schnorr` instead but let me know what you think.

real-or-random commented:
Yes, I think we should change the prefix to `schnorrsig`  too (not just `schnorr`).

Tibo-lg commented:
> Yes, I think we should change the prefix to `schnorrsig` too (not just `schnorr`).

Updated

elichai commented:
We can't use the same SecretKey operations for schnorr secret key. 
as it will not produce the same results as the public key will.

(ie `schnorrsig::PublicKey.add(tweak) != schnorrsig::SecretKey.add(tweak).PublicKey()`)
*I think* we need use the `secp256k1_keypair` API but it doesn't currently expose the seckey back and it means every seckey tweak is also a EC addition

apoelstra commented:
Ah, good catch. I'd forgotten about the keypair API. You're right that we cannot use normal secret keys with the schnorrsig module, at least not without nasty surprises for users.

I think the best thing is to replicate the upstream API ... which does mean that you can't tweak secret keys without also tweaking the public ones. The only alternative is that the user keep track of y-parity themselves and this is too fragile and "hand-rolled crypto"-y for this library.

apoelstra commented:
So @Tibo-lg sorry to jerk you around again, but I think we need to back out the `key.rs.in` stuff, drop the `schnorrsig::SecretKey` type, and work only with keypairs and public keys, like upstream does. Trying to provide an independent `SecretKey` will cause surprising inconsistencies because the keytweaking functions need to be aware of the x-only-key stuff.

Please don't feel obligated to finish this PR btw -- if you drop it, I can pick it up when I find a free moment (though this may not be for a little while).

Tibo-lg commented:
@apoelstra I don't mind updating the PR if you don't mind reviewing it :) 

I removed the `key.rs.in` file, and switched to have `KeyPair` instead of `SecretKey` in the `schnorrsig` module.

I also added `add_assign` to tweak a `KeyPair`, but I'm not really sure on how to test this properly without using a "regular" public key object somewhere. For example, testing `schnorrsig::PublicKey.add(tweak) == schnorrsig::SecretKey.add(tweak).PublicKey()` as @elichai mentioned would require adding an `add` function to `schnorrsig::PublicKey`, but the underlying tweak function returns a "regular" pubkey so I'm not sure how the API should look like. I guess the two options would be to either have two PublicKey types in the schnorrsig module (`XOnlyPublicKey` and `PublicKey`), or to return the parity somehow (but IIUC we want to avoid that?). I'm hoping maybe someone has a better suggestion?

elichai commented:
> I also added `add_assign` to tweak a `KeyPair`, but I'm not really sure on how to test this properly without using a "regular" public key object somewhere. For example, testing `schnorrsig::PublicKey.add(tweak) == schnorrsig::SecretKey.add(tweak).PublicKey()` as @elichai mentioned would require adding an `add` function to `schnorrsig::PublicKey`, but the underlying tweak function returns a "regular" pubkey so I'm not sure how the API should look like. I guess the two options would be to either have two PublicKey types in the schnorrsig module (`XOnlyPublicKey` and `PublicKey`), or to return the parity somehow (but IIUC we want to avoid that?). I'm hoping maybe someone has a better suggestion?


IMO we should only expose the XOnlyPublicKey here, but you can internally convert back and forth between the puckeys where needed, so it will look something like:
```C
secp256k1_xonly_pubkey_tweak_add(ctx, &out_pubkey, &xonlypubkey, &tweak);
secp256k1_xonly_pubkey_from_pubkey(ctx, &xonlypubkey, NULL, &out_pubkey)
```

apoelstra commented:
+1 to @elichai's suggestion, though I think it's reasonable to also expose a function for the user to expose conversion functions, with the understanding that ordinary pubkeys are not useful for any other part of the schnorr API.

Also, I hope we don't wind up converting from xonly to regular anywhere internally, since that's expensive.

elichai commented:
> +1 to @elichai's suggestion, though I think it's reasonable to also expose a function for the user to expose conversion functions, with the understanding that ordinary pubkeys are not useful for any other part of the schnorr API.
> 
> Also, I hope we don't wind up converting from xonly to regular anywhere internally, since that's expensive.

we shouldn't only the opposite (from pubkey to xonly), which is cheap

Tibo-lg commented:
@elichai Thanks for the suggestion. I added `add_assign` to `schnorrsig::PublicKey` and a test for addition.

> I think it's reasonable to also expose a function for the user to expose conversion functions, with the understanding that ordinary pubkeys are not useful for any other part of the schnorr API.

I added `From<::key::PublicKey>` for `schnorrsig::PublicKey` does that fit what you had in mind? Left it in a separate commit for now so it's easier to drop.

apoelstra commented on secp256k1-sys/src/lib.rs:
```diff
@@ -125,6 +139,55 @@ impl Default for Signature {
+
+impl XOnlyPublicKey {
+    /// Create a new (zeroed) x-only public key usable for the FFI interface
+    pub fn new() -> XOnlyPublicKey { XOnlyPublicKey([0; 64]) }
```
We shouldn't have this function, as it creates an invalid public key which will trigger assertation failures in upstream.

apoelstra replied:
I'll fix this in a followup commit/PR. It's a bit subtle because we do need (temporarily) uninitialized values to use the FFI interface.

apoelstra commented on secp256k1-sys/src/lib.rs:
```diff
@@ -125,6 +139,55 @@ impl Default for Signature {
+impl XOnlyPublicKey {
+    /// Create a new (zeroed) x-only public key usable for the FFI interface
+    pub fn new() -> XOnlyPublicKey { XOnlyPublicKey([0; 64]) }
+    pub fn from_array(data: [c_uchar; 64]) -> XOnlyPublicKey {
```
We should rename this to `from_array_unchecked` for similar reasons.

apoelstra commented on secp256k1-sys/src/lib.rs:
```diff
@@ -301,6 +364,92 @@ extern "C" {
+        msg32: *const c_uchar,
+        keypair: *const KeyPair,
+        noncefp: SchnorrNonceFn,
+        noncedata: *const c_void
```
upstream this is a `*mut c_void`

apoelstra commented on src/schnorrsig.rs:
```diff
@@ -0,0 +1,710 @@
+        &mut self,
+        secp: &Secp256k1<C>,
+        tweak: &[u8],
+    ) -> Result<(), Error> {
```
This should return the parity as a bool rather than returning `()`

apoelstra reviewed (approved):
ack 0c937d04f5a198cc6e8687640beec7e3773438c6

Will address nits in a followup PR.
